docs/e2e-testing-headscale-derper-tailscale.md

```diff
@@ -0,0 +1,638 @@
+
+```bash
+IP="$(ssh tailscale-etet-internal-2 -- ip -br address show dev eth0 | awk '{ sub("/.*", "", $3); print $3; }')"
+ssh tailscale-etet-internal-1 -- ssh -o StrictHostKeyChecking=no "$IP" -- hostname
+```
+
+This should also fail, this time with a timeout.
+This is expected, since both machines are on separate private networks, with no routes between them.
+
+```text
+ssh: connect to host 10.1.2.5 port 22: Connection timed out
+```
+
+#### SSH with Tailscale
+
+Bring Tailscale back up:
+
+```bash
+ssh tailscale-etet-internal-1 -- sudo tailscale up
+ssh tailscale-etet-internal-2 -- sudo tailscale up
+```
+
+Attempt to SSH all combinations of machines:
+
+```bash
+ssh tailscale-etet-internal-1 -- $'ssh -o StrictHostKeyChecking=no internal-2 -- echo $(hostname) to \'$(hostname)\''
+ssh tailscale-etet-internal-1 -- $'ssh -o StrictHostKeyChecking=no derper     -- echo $(hostname) to \'$(hostname)\''
+ssh tailscale-etet-internal-2 -- $'ssh -o StrictHostKeyChecking=no internal-1 -- echo $(hostname) to \'$(hostname)\''
+ssh tailscale-etet-internal-2 -- $'ssh -o StrictHostKeyChecking=no derper     -- echo $(hostname) to \'$(hostname)\''
+ssh tailscale-etet-derper     -- $'ssh -o StrictHostKeyChecking=no internal-1 -- echo $(hostname) to \'$(hostname)\''
+ssh tailscale-etet-derper     -- $'ssh -o StrictHostKeyChecking=no internal-2 -- echo $(hostname) to \'$(hostname)\''
+```
+
+All should succeed, and the expected output should be:
+
+```text
+internal-1 to internal-2
+internal-1 to derper
+internal-2 to internal-1
+internal-2 to derper
+derper to internal-1
+derper to internal-2
+```
+
+Check the status of Tailscale:
+
+```bash
+ssh tailscale-etet-internal-1 -- tailscale status
+```
+
+It should now show fresh information about the connection to internal-2:
+
+```text
+100.64.0.1      internal-1           user1        linux   -
+100.64.0.3      derper               user2        linux   -
+100.64.0.2      internal-2           user1        linux   active; direct 52.156.185.242:1024, tx 19372 rx 17988
+```
+
+Note that it shows a direct connection through some public IP address.
+It's unknown what this means, as this is not the DERP relay.
+Tailscale has managed to negotiate some kind of direct connection,
+most likely using the DERP server to help negotiate it.
+
+### Test the DERP server
+
+We can probe the DERP server from Tailscale,
+to check it's configured and operating as expected.
+
+First, we can use the `tailscale netcheck` subcommand:
+
+```bash
+ssh tailscale-etet-internal-1 -- tailscale netcheck
+```
+
+This will output several lines,
+but we're interested in the last few about DERP.
+The output should look like this:
+
+```text
+Report:
+        ...
+        * Nearest DERP: Custom Region One
+        * DERP latency:
+                - one: 1.1ms   (Custom Region One)
+```
+
+TODO: figure out what firewall rules to set to force traffic through DERP, so we can test ssh'ing to a machine through the relayed connection
```

Comment:
I tried playing with ufw for the following two approaches:


```bash
ssh tailscale-etet-internal-1 -- <<'EOF'
set -x
yes | sudo ufw enable
sudo ufw default deny incoming
sudo ufw default deny outgoing
sudo ufw allow 22
sudo ufw allow out from any to any port 443 proto tcp
sudo ufw allow out from any to any port 3478 proto udp
sudo ufw allow out from any to any port 53
sudo ufw status
EOF

ssh tailscale-etet-internal-1 -- <<'EOF'
set -x
yes | sudo ufw enable
sudo ufw default allow
sudo ufw default allow outgoing
sudo ufw deny out from any to any port 41641 proto udp
sudo ufw status
EOF

```

The first appeared to force tailscale to use the derp relay,
however, I couldn't ssh between the nodes.

I think the second didn't force use of the derp relay.

Reply:
Ah, it turns out that blocking all outgoing UDP traffic does the job. :)